### PR TITLE
Port unbind_to

### PR DESCRIPTION
**Pull request by nickdrozd**

### CODE REVIEW

File: src/eval.c [hunk 1]
```diff
@@ -2605,36 +2605,6 @@ set_unwind_protect_ptr (ptrdiff_t count, void (*func) (void *), void *arg)
...
-
-Lisp_Object
-unbind_to (ptrdiff_t count, Lisp_Object value)
-{
-  Lisp_Object quitf = Vquit_flag;
-
-  Vquit_flag = Qnil;
-
-  while (specpdl_ptr != specpdl + count)
-    {
```
[the remark this answers is not in the document]
Comment (nickdrozd) [reply src/eval.c]: fixed

File: rust_src/src/eval.rs [hunk 2]
```diff
@@ -1105,4 +1108,32 @@ pub fn funcall(args: &mut [LispObject]) -> LispObject {
...
+    unsafe {
+        while current_thread.m_specpdl_ptr != current_thread.m_specpdl.offset(count) {
+            current_thread.m_specpdl_ptr = current_thread.m_specpdl_ptr.offset(-1);
+            let this_binding = current_thread.m_specpdl_ptr;
+
+            do_one_unbind(this_binding, true, Set_Internal_Bind::SET_INTERNAL_UNBIND);
+        }
+    }
+
+    unsafe {
```
[the remark this answers is not in the document]
Comment (nickdrozd) [reply rust_src/src/eval.rs]: I actually thought the blocks were too broad, so I added more.

File: rust_src/src/eval.rs [hunk 3]
```diff
@@ -1105,4 +1108,35 @@ pub fn funcall(args: &mut [LispObject]) -> LispObject {
...
+        // bindings are made during some of the code we run.
+
+        current_thread.m_specpdl_ptr = unsafe { current_thread.m_specpdl_ptr.offset(-1) };
+
+        let this_binding = current_thread.m_specpdl_ptr;
+
+        unsafe { do_one_unbind(this_binding, true, Set_Internal_Bind::SET_INTERNAL_UNBIND) };
+    }
+
+    if unsafe { globals.Vquit_flag }.is_nil() && quitf.is_not_nil() {
```
Comment (shaleh) [inline rust_src/src/eval.rs]:
```suggestion
    unsafe {
        if globals.Vquit_flag.is_nil() && quitf.is_not_nil() {
            globals.Vquit_flag = quitf;
        }
    }
```

File: rust_src/src/eval.rs [hunk 4]
```diff
@@ -1105,4 +1108,35 @@ pub fn funcall(args: &mut [LispObject]) -> LispObject {
     val
 }
 
+/// Pop and execute entries from the unwind-protect stack until the
+/// depth COUNT is reached. Return VALUE.
+#[no_mangle]
+pub extern "C" fn unbind_to(count: libc::ptrdiff_t, value: LispObject) -> LispObject {
+    let mut current_thread = ThreadState::current_thread();
+
```
Comment (shaleh) [inline rust_src/src/eval.rs]: This whole function is unsafe basically. We should just embrace it and wrap it in a block.

Comment (nickdrozd) [reply rust_src/src/eval.rs]: Fair enough, fixed

**Comment by shaleh** [comment]
@harryfei I reverting a merge mistake. I'd like to hold off on merges until they are done.

**Comment by nickdrozd** [comment]
@shaleh There was a merge conflict. Rebased on master and trying again.

**Comment by shaleh** [comment]
Looks go to merge once the tests complete.

**Pull request merged**